lib/jekyll/renderer.rb

```diff
@@ -169,12 +169,16 @@ def place_in_layouts(content, payload, info)
     # Returns nothing
     private
     def validate_layout(layout)
-      return unless invalid_layout?(layout)
-      Jekyll.logger.warn(
-        "Build Warning:",
-        "Layout '#{document.data["layout"]}' requested "\
-        "in #{document.relative_path} does not exist."
-      )
+      if invalid_layout?(layout)
+        Jekyll.logger.warn(
+          "Build Warning:",
+          "Layout '#{document.data["layout"]}' requested "\
+          "in #{document.relative_path} does not exist."
+        )
+      elsif !layout.nil?
+        layout_source = layout.path.start_with?(site.source) ? :site : :theme
```

Comment:
What do you think about adding this method to the `Layout` class, so you could ask the layout where it's from? 😄

Reply:
I was planning to open a PR after this merged to do just what you asked.. (but very different from what's here), I was aiming to make most minimal changes to what @benbalter had originally authored..